### PR TITLE
Docs

### PR DESCRIPTION
**Pull request by ramarag**
skip ci please

@gkhanna79 @eerhardt @natemcmaster PTAL

### CODE REVIEW

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+The layout within `netcoreapp*` folders is a NuGet cache layout.
+
++ **User**: The user local location can be used to override the global location when probing assemblies and uses a very similar layout, except that the `packages/install` folder is NOT supported in the user profile. It is only supported in the global location.
+
+```
+    - %USERPROFILE% or $HOME/
+        - .dotnet
+          - packages 
+            - {arch=x86|x64}
+                = do =
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: I don't understand this line.

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+```
+
+**OPEN**: Should we have a `dotnet` relative location as well, if operating out of a zip installation of .NET Core binaries?
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: Yes.

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        "version": "2.0.0",
+        "name": "Microsoft.NETCore.App"
+    }
+}
+```
+Note that this is different from current behavior of `dotnet run` for an application that specifies the `<RuntimeIdentifier/>` tag in the csproj.
+**Current Behavior:** Picks `M.N.A` assemblies out of the NuGet cache without taking advantage of optimizations available from the shared `Microsoft.NETCore.App`.
+**New Behavior:** Picks `M.N.A` assemblies from the shared framework and the rest of them from the shared package store or the NuGet cache.
+
+`dotnet build` can take advantage of the `refs` folder available at the `packages/install/` folder from the `dotnet` root directory enabling the offline-restore-build scenario. Although we are designing to augment `dotnet build` in the future regarding the reference assemblies, for the scope of this work we'll focus only on runtime assemblies.
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: Should we be discussing `refs` at all in this doc since this work is out of scope?

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+The layout is composed from a list of package names and versions specified as xml: 
+
+**Roslyn Example**
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+ <ItemGroup>
+    <PackageReference Include="Microsoft.CodeAnalysis" Version="1.3.2" />
+    <PackageReference Include="Microsoft.CodeAnalysis.VisualBasic" Version="1.3.2" />
+    <PackageReference Include=""Microsoft.CodeAnalysis.VisualBasic.Features" Version="1.3.2" />
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: (nit) double `""` after the `Include=`

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+  </ItemGroup>
+</Project>
+```
+
+
+and issue a command like below:
+
+```
+dotnet cache --entries packages.xml --framework netcoreapp2.0 [--output C:\Foo] --runtime win7-x64 --framework-version 2.0.0-preview2-00001 [--no-optimize]
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: Can these values come from the `packages.xml` file?  That way I don't have to specify them on the command line every time?

ex.
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp2.0</TargetFramework>
```

then I don't have to specify `--framework` on the command line.  Same for `--runtime` and `--framework-version`

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+--framework-version   The Microsoft.NETCore.App package version that will be used to run the assemblies
+
+```
+NOTE: It is a requirement that `packages.xml` is of msbuild format, because it forms the entry point from which the rest of the SDK's functionality can be accessed
+
+Hosting providers would create a `packages.xml` file corresponding to the packages that will be shared in their hosting environment and specify the file to `dotnet cache`. The file can be on the file system or from an URL. The TFM argument is used in the shared package layout described above.
+
+If `--optimize` is specified, we would precompile all the managed assets to native code in a temp folder before copying to the output folder. If crossgen is used, it would be the one acquired in the closure of the `Microsoft.NETCore.App` specified by the `--framework-version` option. Also, if no `--output` folder is specified, then the default is `~/.dotnet` or `%USERPROFILE%\.dotnet\`. The output asset files will be present in the following layout: `$HOME/.dotnet/packages/{tfm}/{package-name}/{package-version}/{asset-path}`.
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: There is no `--optimize` option above.  Instead there is a `--skip-optimization`

Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: > The output asset files will be present in the following layout

The output asset files should be in the same layout as the host expects above.  That way I can invoke `dotnet cache`, and without moving assets around, they are picked up when I `dotnet run`.

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+The output folder will be consumed by the runtime by adding to the `DOTNET_SHARED_PACKAGES_ROOTS` environment variable. See probe precedence below.
+
+# Building apps with shared packages
+
+The current mechanism to build applications that share assemblies is by not specifying a RID in the project file. Then, a portable app model is assumed and assemblies that are part of Microsoft.NETCore.App are found under the `dotnet` install root. With shared package store, applications have the ability to filter any set of packages from their publish output. Thus the decision of a portable or a standalone application is not made at the time of project authoring but is instead done at publish time.
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: Can a user publish for standalone *and* filter some assemblies at the same time?  For example, filter out Roslyn, but put the .NET runtime assemblies in my publish directory.  Does this scenario make sense?

Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: Oh, I think the next paragraph takes care of that... It says this should be an error.

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+```
+    - dotnet.exe
+    - shared
+        - netcoreapp2.0
+            + 2.0.0-preview2-00001
+    - packages
+        - install
+            + refs
+            + netcoreapp2.0
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: How do we expect users to handle pre-release versions?  For example, we ship `2.0.0-preview1`, and users start using it to create caches.  Now in `2.0.0-preview2` we make some crossgen bug fixes.  When the user upgrades to `2.0.0-preview2`, all their old, cached assemblies will still be under the `netcoreapp2.0` folder.  Do we expect them to delete these folders out whenever they move across pre-release versions?

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
++ `$CORE_SERVICING` on Unix or `%ProgramFiles(x86)%\coreservicing` on Windows.
++ `DOTNET_SHARED_PACKAGES_ROOTS` in the chained order
++ The user local shared package store in `%USERPROFILE%\.dotnet\packages` or `$HOME/.dotnet/`
++ The global shared package store
++ The Shared Framework directory
++ ~~`DOTNET_HOSTING_OPTIMIZATION_CACHE` deprecated in favor of `DOTNET_SHARED_PACKAGES_ROOTS`~~
++ `--additionalprobingpaths` specified in the command line
++ `runtimeOptions.additionalProbingPaths` (includes NuGet cache probe specified by the CLI for `dotnet run`)
++ Application `bin` directory
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: This is not what @gkhanna79  mentioned last week - with "the app always wins".

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+
+Publish will be enhanced to support a filter profile file specified as xml. This file explicitly lists all asset packages that need to be trimmed out of the publish output. The following are examples of how various application types can be published.
+
+Publish a portable app:
+
+```dotnet publish```
+
+Publish a standalone app for the current RID:
+
+```dotnet publish --standalone```
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: There is no `--standalone` in the current CLI.

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+Publish a standalone app for win7-x64 filtering out the publish profile:
+
+```dotnet publish --runtime win7-x64 filter https://asp.net/core/dev/1.2.0/profile.xml```
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: `filter` should have a `--` in front of it.

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+
+Publishing a portable app for win7-x64 filtering out the publish profile:
+
+```dotnet publish filter https://asp.net/core/dev/1.2.0/profile.xml```
+
+Publishing a Windows level portable application (a.k.a rid-specific portable app) i.e., filter all non-windows RIDs but split the bin directory on windows RIDs (win8, win7 and win10, for example):
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: This scenario doesn't exist today.

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+    - dotnet cache
+    - Zip the layout
+    - Deploy on hosting servers
+    - Publish apps with filter
+
+## Work Items
+
+### Core-Setup
+
++ Change name of `Microsoft.NETCore.App` to `netcoreapp2.0` in the dotnet root. Note that this has to maintain compat.
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: Is this still happening?

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
++ Move the Roslyn assemblies out of the shared framework.
++ Build installers on Windows, Ubuntu and OSX for the Roslyn (shared) packages.
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: We aren't doing this anymore.  Roslyn will be included in the ASP.NET runtime installer.

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
++ Implement host probing for the user and global package stores.
++ Incorporate changes in host for the TFM changes.
+
+### CLI
+
++ Make CLI consume Roslyn dependencies out of the shared packages root as these are not part of M.N.A anymore.
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: CLI doesn't use Roslyn from M.N.A since it requires `v2.0.0` of Roslyn and M.N.A only has `v1.3.0`.

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
++ Make `dotnet restore` restore as though project is `type: platform`.
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: Not sure I understand this one.  `type: platform` no longer exists at all in NuGet.

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
++ Make `dotnet build` treat projects as though they are `type: platform`.
```
Comment (eerhardt) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: I don't think this is necessary anymore.

File: Documentation/design-docs/corehost.md [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+* `DOTNET_PACKAGES_CACHE` - The secondary cache. This is used by shared hosts (such as Azure) to provide a cache of pre-downloaded common packages on a faster disk. If not set, it is not used.
+
+Given the Package ID, Package Version, Package Hash and Asset Relative Path provided in the runtime configuration file, **and the assembly is not serviced** (see the full resolution algorithm below) resolution proceeds as follows (Unix-style paths will be used for convenience but these variables and paths all apply to Windows as well):
+
+1. If `DOTNET_PACKAGES_CACHE` is non-empty, read the file `$DOTNET_PACKAGES_CACHE/[Package ID]/[Package Version]/[Package Id].[Package Version].nupkg.sha512` if present. If the file is present and the content matches the `[Package Hash]` value from the dependencies file. Use that location as the Package Root and go to 3
```
Comment (eerhardt) [inline Documentation/design-docs/corehost.md]: I thought the check for the `.nupkg.sha512` file was removed in the 2.0 code.

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md (hunk 5 above)
Comment (ramarag) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: yes we will honor the msbuild way of specifying them. but command line will win over the specified properties

File: Documentation/design-docs/DotNetCore-SharedPackageStore.md (hunk 9 above)
Comment (weshaggard) [inline Documentation/design-docs/DotNetCore-SharedPackageStore.md]: Yes I was just about to post a similar question.
